profit/config.py

```diff
@@ -146,126 +255,333 @@ def from_file(cls, filename='profit.yaml'):
-        else:
-            self['config_path'] = path.abspath(path.join(getcwd(), filename))
 
-        """ Variable configuration
-        kind: Independent, Uniform, etc.
-        range: (start, end, step=1) or {'dependent variable': (start, end, step=1)} for output
-        dtype: float64
-        """
+@BaseConfig.register("run")
+class RunConfig(AbstractConfig):
+    """Run configuration with the following sub classes:
+        - runner
+            - local
+            - slurm
+        - interface
+            - memmap
+            - zeromq
+        - pre
+            - template
+        - post
+            - json
+            - numpytxt
+            - hdf5
+
+    A default sub class which just updates the entries from a user input is also implemented and used if the
+    class from the user input is not found.
+
+    Custom config classes can also be registered, e.g. as a custom runner:
+
+    .. code-block:: python
+
+        @RunnerConfig.register("custom")
+        class CustomRunner(LocalRunnerConfig):
+            def process_entries(self, base_config):
+                # do something else than the usual LocalRunnerConfig
+                pass
+
+    Default values from the global profit.defaults.py file are loaded.
+    """
+    _sub_configs = {}
+
+    def __init__(self, **entries):
+        self.set_defaults(defaults.run)
+        self.update(**entries)
+
+        for key, sub_config in self._sub_configs.items():
+            attr = getattr(self, key.lower())
+            if isinstance(attr, str):
+                attr = {'class': attr}
+            try:
+                setattr(self, key.lower(), sub_config._sub_configs[attr['class']](**attr))
+            except KeyError:
+                setattr(self, key.lower(), sub_config._sub_configs['default'](**attr))
+
+    def process_entries(self, base_config):
+        """Set 'include' and paths and process entries of sub configs."""
+        from profit.util import load_includes
+
+        if isinstance(self.include, str):
+            self.include = [self.include]
+
+        for p, include_path in enumerate(self.include):
+            if not path.isabs(include_path):
+                self.include[p] = path.abspath(path.join(base_config.base_dir, include_path))
+        load_includes(self.include)
+
+        if not path.isabs(self.log_path):
+            self.log_path = path.abspath(path.join(base_config.base_dir, self.log_path))
+
+        for key in self._sub_configs:
+            getattr(self, key.lower()).process_entries(base_config)
+
+
+@RunConfig.register("runner")
+class RunnerConfig(AbstractConfig):
+    """Base Runner config."""
+    _sub_configs = {}
+
+    def __init__(self, **entries):
+        for key, value in self._sub_configs.items():
+            if value.__name__ == self.__class__.__name__:
+                self.set_defaults(getattr(defaults, f"run_runner_{key}"))
+                self.update(**entries)
+                break
+
+
+@RunnerConfig.register("local")
```

Comment:
I would personally prefer it if the actual configurations are next to the relevant implementation. This would ensure that all relevant information of that component is just in one file.
But that's probably a discussion for another time...